modules/deriving/src/main/scala/shapeless3/deriving/kinds.scala

```diff
@@ -166,28 +158,27 @@ object K0:
     inline def widen[G[t] >: F[t]]: CoproductInstances[G, T] =
       inst.asInstanceOf
 
-  inline given mkInstances[F[_], T](using gen: Generic[T]): Instances[F, T] =
+  inline given mkInstances[F[_], T](using gen: Mirror.Of[T]): Instances[F, T] =
     inline gen match
       case p: ProductGeneric[T] => mkProductInstances[F, T](using p)
       case c: CoproductGeneric[T] => mkCoproductInstances[F, T](using c)
 
-  inline given mkProductInstances[F[_], T](using gen: ProductGeneric[T]): ProductInstances[F, T] =
+  inline given mkProductInstances[F[_], T](using gen: Mirror.ProductOf[T]): ProductInstances[F, T] =
     ErasedProductInstances[K0.type, F[T], LiftP[F, gen.MirroredElemTypes]](gen)
 
-  inline given mkCoproductInstances[F[_], T](using gen: CoproductGeneric[T]): CoproductInstances[F, T] =
+  inline given mkCoproductInstances[F[_], T](using gen: Mirror.SumOf[T]): CoproductInstances[F, T] =
     ErasedCoproductInstances[K0.type, F[T], LiftP[F, gen.MirroredElemTypes]](gen): CoproductInstances[F, T]
 
 object K1:
-  type Kind[C, O[_]] = C {
-    type Kind = K1.type
+  type Refl[M, O[_]] = M {
```

Comment:
What's the idea behind the name `Refl` here? Makes me think of `<:<.refl` and the likes. I think `Generic` was easier to understand for users (previously `mkInstances` mentioned only `Generic[O]` instead of `Kind[Mirror, O]`.

Reply:
`Refl` like reflection. Dunno, I also don't like the name `Kind` given that it also has a type member `Kind`. So what's the meaning of `Kind#Kind`? `Generic` is easier to understand but it has the problem you reported that custom mirrors don't work with `Generic`. Whereas they do work with `Refl`.

Reply:
I agree that `Kind` is not a great name either. How about something like `type FixMirrorKind[M <: Mirror, O[_]]`? Also it should be mentioned that changing the name, as well as removing `type Kind` is technically a breaking change in public API, though unlikely to cause much trouble.

Reply:
But then what does `FixMirrorKind[Mirror, Option]` mean? Btw, adding a bound `M <: Mirror` would be a binary-breaking change, whereas changing the type alias is only source-breaking (if anyone is using it at all, I struggle to think of a use-case for it outside the library). I guess I could add back the `Kind` type alias and it would be also source-compatible.

Reply:
Yeah, it doesn't read well, but it's still more meaningful, don't you think? 

`(Mirror WithKind K0)[Option]` would be nice.
```scala 3
sealed trait Kinds
object K1 extends Kinds {
  ...
}

infix type WithKind[M <: Mirror, K <: Kinds] = K match {
    case K1 => [O[_]] =>> M {
       type MirroredType[X] = O[X]
       ...
    }
    ... 
}
```
Probably binary incompatible as well. I'm just brainstorming. But it's only a name, so no reason to bikeshed.

Reply:
Infix type sounds interesting, perhaps `Mirror of O`, `Mirror.Product of O` and `Mirror.Sum of O`

Reply:
Hmm, I think `M <: Mirror` was giving me mima errors because I didn't clean, it might be actually ok.

Reply:
We already have `Mirror.Of[O]` though. Could be confusing, too.

Reply:
Not for higher kinds, we don't have it

Reply:
The name is very similar, I mean.

Reply:
It serves the same purpose